### PR TITLE
Binary to decimal

### PR DESCRIPTION
**Pull request by AlhassenSabeeh**
---
- #46 
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Adamx090 — CHANGES_REQUESTED** [review]

File: solutions/binary_to_decimal.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    2796202
+    """
+    # The binary input should be a string containing only '0' or '1' characters
+    assert isinstance(binary, str), "Binary is not a string"
+    assert set(binary) <= {"0", "1"}, "Binary string contains invalid characters"
+    # The strategy recursively converts a binary string to its decimal value
+    # by processing each digit and reducing the string until empty
+    if len(binary) == 0:
+        # Base Case,  f(ϵ)=0
+        return 0
```
[the remark this answers is not in the document]
Comment (AlhassenSabeeh) [reply solutions/binary_to_decimal.py]: Amazing @Adamx090, 
Nice catch, this is one of the hardest things in writing and reviewing code, "off-by-one" error.
I have already fixed the error and made what you suggested to modify it, which you can see the changes in the last commit.

File: solutions/binary_to_decimal.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    # Recursive Case, f(b1b2…bn) = b1 . 2^(n−1) + f(b2b3…b<n-1>)
+    #      /Return Step/                         /Recursive Case/ /Reduction Step/
```
[the remark this answers is not in the document]
Comment (AlhassenSabeeh) [reply solutions/binary_to_decimal.py:52]: In this thread, I stopped and thought about it when I wrote this piece of the code and used the recursion strategy as a reflection of the mathematical model, which I wrote inside the code as comments, and then wrote some comments before it to illustrate this model.
That is why I broke that rule.
I am not sure if I remove those comments, is the mathematical model enough to explain the strategy?

[the remark this answers is not in the document]
Comment (Adamx090) [reply solutions/binary_to_decimal.py:52]: Actually, now that you mentioned it, those comments helped me follow the code well. I think you are correct here, specially since recursion can be tricky, and it requires explanation.

File: solutions/binary_to_decimal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Adamx090) [reply solutions/binary_to_decimal.py]: Thanks for reminding me that this was a "off-by-one" error, I didn't consider it earlier.
Everything looks great! I will merge it now.

**Review by Adamx090 — APPROVED** [review]

File: solutions/binary_to_decimal.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Adamx090) [reply solutions/binary_to_decimal.py:52]: After a second look, it looks like there is an equal amount of comments and code.
